src/crimson/osd/osd.cc

```diff
@@ -1180,4 +1199,28 @@ blocking_future<Ref<PG>> OSD::wait_for_pg(
   return pg_map.get_pg(pgid).first;
 }
 
+seastar::future<> OSD::prepare_to_stop()
+{
+  if (osdmap && osdmap->is_up(whoami)) {
+    state.set_prestop();
+    return seastar::with_timeout(
+      seastar::timer<>::clock::now() + std::chrono::seconds(5),
+      monc->send_message(
+	  make_message<MOSDMarkMeDown>(
+	    monc->get_fsid(),
+	    whoami,
+	    osdmap->get_addrs(whoami),
+	    osdmap->get_epoch(),
+	    true)).then([this] { return stop_acked.get_future(); })
+    ).handle_exception([this](auto&& eptr) {
+      if (state.is_prestop()) {
```

Comment:
i think a better and probably simpler implementation is something like:
```c++
  if (osdmap && osdmap->is_up(whoami)) {
    state.set_prestop();
    return monc->send_message(
      make_message<MOSDMarkMeDown>(
            monc->get_fsid(),
            whoami,
            osdmap->get_addrs(whoami),
            osdmap->get_epoch(),
            true)).then([this] {
    const auto timeout =
      std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::duration<double>(
          local_conf().get_val<double>("osd_mon_shutdown_timeout")));
    return seastar::with_timeout(
      seastar::steady_clock_type::now() + timeout,
      stop_acked.get_future());
    }).handle_exception_type([](seastar::timed_out_error&) {
      // ignore
    });
  } else {
    return seastar::now();
  }
```

in which,

1. move `send_message()` call out of timeout, as we should not take the time for sending a message into the timeout.
1. `assert(state.is_prestop())` as prestop should always set at that moment
1. move `state.set_stopping();` into the `else` branch for better readability
1. check for `seastar::timed_out_error` explicitly.

also, i think there is no need to call `state.set_stopping();` in `got_stop_ack()`. probably a better place for calling `state.set_stopping();` is `OSD::stop()`. like:
```c++
seastar::future<> OSD::stop()
{
  logger().info("stop");
  // see also OSD::shutdown()
  return prepare_to_stop().then([this] {
    state.set_stopping();
    logger().debug("prepared to stop");
```